### PR TITLE
fix: Performance improvement for scattergl with many points. Issue #7065

### PR DESCRIPTION
**Pull request by giuseppe-straziota**
The performance issues are always intriguing, and I tried to determine why the browser freezes when handling a large number of points. I believe the problem stems from two main factors. 

The first factor is the hover effect, which triggers a function that calculates the distance between the mouse pointer and each point in the set. This process can certainly contribute to the performance problem, I believe the best improvement would be using a filter calculated by a window determined by the pointer coordinates with a delta, in this way, we can calculate the distance only for a subset. One possible improvement could be using squared distances for comparison, thereby avoiding the computation of square roots, similar to the method utilized by the [K-d tree algorithm](https://en.wikipedia.org/wiki/K-d_tree), if the plot allows it.

The second factor involves a loop in the part of the code I modified, where "newDistance" is created for every point. This results in excessive overhead for the garbage collector, causing the browser to freeze.
```
for(var i = 0; i < cd.length; i++) {
            var newDistance = distfn(cd[i]);
            if(newDistance <= pointData.distance) {
                pointData.index = i;
                pointData.distance = newDistance;
            }
        }
```
I moved the variable outside the cycle and used a precalculated value for the array length. I conducted tests in the development environments provided by Plotly using the test_dashboard tool.
The values represent average times measured in milliseconds..
 
| | 6000 | 10000 | 100000 | 1Milion | 10Milion | 20Milion |
| :---         |     :---:      |     :---:      |     :---:      |     :---:      |     :---:      |          ---: |
| inside for   | 0,841015625 | 1,892871094 | 17,34858398 | 154,4843262 | 1412,1573 | 2733,622534 |
| outside for   | 0,7019287109 | 1,51418457 | 13,50952148 | 113,6539307 | 1057,211401 | 1078,00 |

The red line shows the performance for the variable inside, the blue line outside. The vertical axis (Y-axis) represents the time measured in milliseconds.
image

Of course, the performance depends on the engine and the environment in which it runs, but I think this conveys the idea of performance.

Please let me know if this can be the first step to improving performance.

### CODE REVIEW

**Comment by giuseppe-straziota** [comment]
I apologize, everyone; I mistakenly referenced the wrong issue. The correct issue number is #7065 sorry.

File: draftlogs/7301_fix.md [hunk 1]
```diff
@@ -0,0 +1 @@
+- Fix: Performance improvement for scattergl with many points [[#7301](https://github.com/plotly/plotly.js/pull/7301)]
```
Comment (archmoj) [inline draftlogs/7301_fix.md:1]:
```suggestion
- Performance improvement for scattergl traces with many points [[#7301](https://github.com/plotly/plotly.js/pull/7301)],
   with thanks to @giuseppe-straziota for the contribution!
```

**Review by archmoj — APPROVED** [review]
Excellent PR. :trophy: 
Many many dancers: 
:dancer: :dancer: :dancer: :dancer: :dancer: :dancer: :dancer: :dancer: :dancer: :dancer: :dancer: :dancer: :dancer: :dancer: :dancer: :dancer: :dancer: :dancer: :dancer: :dancer: :dancer: :dancer: :dancer: :dancer: :dancer: :dancer: :dancer: :dancer: :dancer: :dancer: :dancer: :dancer: :dancer: :dancer: :dancer: :dancer: :dancer: :dancer: :dancer: :dancer: :dancer: :dancer: :dancer: :dancer: :dancer: :dancer: :dancer: :dancer: :dancer: :dancer: :dancer: :dancer: :dancer: :dancer: :dancer: :dancer: :dancer: :dancer: :dancer: :dancer: :dancer: :dancer: :dancer: :dancer: :dancer: :dancer: :dancer: :dancer: :dancer: :dancer: :dancer: :dancer: :dancer: :dancer: :dancer: :dancer: :dancer: :dancer: :dancer: :dancer: :dancer: :dancer: :dancer: :dancer: :dancer: :dancer: :dancer: :dancer: :dancer: :dancer: :dancer: :dancer: :dancer: :dancer: :dancer: :dancer: :dancer: :dancer: :dancer: :dancer: :dancer: :dancer: :dancer: :dancer: :dancer: :dancer: :dancer: :dancer: :dancer: :dancer: :dancer: :dancer: :dancer: :dancer:

**Comment by giuseppe-straziota** [comment]
I'm glad to have been helpful :)

**Pull request merged**